grudge/op.py

```diff
@@ -179,77 +197,353 @@ def _single_axis_derivative_kernel(
+                        tagged=(FirstAxisIsElementsTag(),
+                                OutputIsTensorProductDOFArrayOrdered(),),
+                        arg_names=("diff_mat", f"vec_{xyz_axis}")
+                    )
+                )
+
+        grad = actx.np.stack(grad)
+        return tag_axes(
+            actx,
+            {
+                0: DiscretizationAmbientDimAxisTag(),
+                1: DiscretizationElementAxisTag(),
+                2: DiscretizationDOFAxisTag()
+            },
+            actx.einsum(
+            "xrej,rej->xej",
+            ijm,
+            grad,
+            tagged=(FirstAxisIsElementsTag(),),
+            arg_names=("inv_jac_t", f"vec")
+        ))
+
+    # }}}
+
+
+    # {{{ simplicial grad
+
+    def compute_simplicial_grad(actx, in_grp, out_grp, get_diff_mat, vec_i,
+                                ijm_i, metric_in_matvec):
+        return actx.einsum(
+            "xrej,rij,ej->xei" if metric_in_matvec else "xrei,rij,ej->xei",
+            ijm_i,
+            get_diff_mat(
+                actx,
+                out_element_group=out_grp,
+                in_element_group=in_grp
+            ),
+            vec_i,
+            arg_names=("inv_jac_t", "ref_stiffT_mat", "vec"),
+            tagged=(FirstAxisIsElementsTag(),))
+
+    # }}}
+
+
     per_group_grads = [
-        # r for rst axis
-        # x for xyz axis
-        actx.einsum("xrej,rij,ej->xei" if metric_in_matvec else "xrei,rij,ej->xei",
-                    ijm_i,
-                    get_diff_mat(
-                        actx,
-                        out_element_group=out_grp,
-                        in_element_group=in_grp
-                    ),
-                    vec_i,
-                    arg_names=("inv_jac_t", "ref_stiffT_mat", "vec"),
-                    tagged=(FirstAxisIsElementsTag(),))
+        compute_tensor_product_grad(actx, in_grp, get_diff_mat, vec_i, ijm_i,
+                                    metric_in_matvec)
+        if isinstance(in_grp, TensorProductElementGroup)
+        else compute_simplicial_grad(actx, in_grp, out_grp, get_diff_mat, vec_i,
+                                     ijm_i, metric_in_matvec)
+
         for out_grp, in_grp, vec_i, ijm_i in zip(
             out_discr.groups, in_discr.groups, vec,
-            inv_jac_mat)]
+            inv_jac_mat)
+    ]
 
     return make_obj_array([
             DOFArray(
                 actx, data=tuple([pgg_i[xyz_axis] for pgg_i in per_group_grads]))
             for xyz_axis in range(out_discr.ambient_dim)])
 
+# }}}
+
+
+# {{{ divergence
 
 def _divergence_kernel(actx, out_discr, in_discr, get_diff_mat, inv_jac_mat, vec,
         *, metric_in_matvec):
     # See _single_axis_derivative_kernel for comments on the usage scenarios
     # (both strong and weak derivative) and their differences.
+
+
+    # {{{ tensor product div
+
+    def compute_tensor_product_div(actx, grp, diff_mat, vec, ijm):
```

Comment:
The result of the simplicial version of these operators lives on the the base discretization.  The TPE versions produce results that live on the quadrature discretization, preventing overintegration from working as expected.

Reply:
Are you referring to the use of `in_grp` and `out_grp` between the two versions?

Reply:
> Are you referring to the use of `in_grp` and `out_grp` between the two versions?

That's definitely a code difference I can see, but I'm actually referring to the observation that if you send in *vec* that is on the quadrature discretization, this routine produces a result that lives on the quadrature discretization instead of one that lives on the base discretization as the simplicial version does.

Reply:
>if you send in vec that is on the quadrature discretization, this routine >produces a result that lives on the quadrature discretization instead of one >that lives on the base discretization

I think the reason this happens is because the routine only considers `in_grp`. I'm not sure why I wrote this to only consider `in_grp`. I *think* I remember some odd behavior (or outright errors) when grabbing the 1D operators.

Reply:
There seems to be only a TPE case in this routine for when in_group == out_group:

https://github.com/a-alveyblanc/grudge/blob/a662445c43d38d97324cb9744d144e0fa9a47a42/grudge/op.py#L755

Reply:
I remember writing it that way, but I can't remember the exact reason. I'll look into it a bit and report back